### PR TITLE
CFE-3088: New stat function, ProtocolStat

### PR DESCRIPTION
**Pull request by karlhto**
A simpler, non-cached version of `cf_remote_stat`. Might need some more proofing, as well as maybe supporting links.

### CODE REVIEW

File: libcfnet/stat_cache.c [hunk 1]
```diff
@@ -212,55 +212,24 @@ int cf_remote_stat(AgentConnection *conn, bool encrypt, const char *file,
...
-               &d1, &d2, &d3, &d4, &d5, &d6, &d7, &d8, &d9, &d10, &d11, &d12, &d13);
-
-        if (ret < 13)
+        int ret = StatParseResponse(recvbuffer + strlen("OK: "), &cfst);
+        if (ret == -1)
         {
-            Log(LOG_LEVEL_ERR, "Cannot read SYNCH reply from '%s', only %d/13 items parsed", conn->remoteip, ret );
+            Log(LOG_LEVEL_ERR,
+                "Cannot read SYNCH reply from '%s', only %d/13 items parsed",
+                conn->remoteip, ret );
```
Comment (vpodzime) [inline libcfnet/stat_cache.c]: Extra space after `ret`.

File: libcfnet/stat_cache.c [hunk 2]
```diff
@@ -349,3 +304,85 @@ const Stat *StatCacheLookup(const AgentConnection *conn, const char *file_name,
...
+    statbuf->cf_size = (off_t) d6;
+    statbuf->cf_atime = (time_t) d7;
+    statbuf->cf_mtime = (time_t) d8;
+    statbuf->cf_ctime = (time_t) d9;
+    statbuf->cf_makeholes = (char) d10;
+    statbuf->cf_ino = d11;
+    statbuf->cf_nlink = d12;
+    statbuf->cf_dev = (dev_t)d13;
+
+    return 0;
```
Comment (vpodzime) [inline libcfnet/stat_cache.c]: Should be `return ret`, right?

File: libcfnet/protocol.c [hunk 3]
```diff
@@ -88,3 +89,93 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
...
+    time_t tloc = time(NULL);
+    if (tloc == (time_t) -1)
+    {
+        Log(LOG_LEVEL_ERR, "Couldn't read system clock (time: %s)",
+            GetErrorStr());
+        tloc = 0;
+    }
+
+    char buf[CF_BUFSIZE] = {0};
+    // Won't this cause problems on 32-bit systems? time_t would always be 64-bit, right?
```
Comment (vpodzime) [inline libcfnet/protocol.c]: I don't think this is a safe assumption.

File: libcfnet/protocol.c [hunk 4]
```diff
@@ -88,3 +89,93 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
...
+    }
+
+    Stat cf_stat;
+    ret = StatParseResponse(buf, &cf_stat);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to parse the response from the server",
+            __func__);
+        return -1;
```
Comment (vpodzime) [inline libcfnet/protocol.c]: I don't think printing function names in `ERR` messages which are shown to the user by default is a good idea. It's good for `DEBUG` messages, but I think it only adds confusion to `ERR` messages.

File: libcfnet/protocol.h [hunk 5]
```diff
@@ -64,4 +64,17 @@
  */
 Seq *ProtocolOpenDir(AgentConnection *conn, const char *path);
 
+/**
+ * Receives statistics about a remote file.
+ *
+ * For the time being this is only a wrapper for #cf_remote_stat
```
Comment (vpodzime) [inline libcfnet/protocol.h]: Is it?

File: libcfnet/protocol.c [hunk 6]
```diff
@@ -88,3 +89,93 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
...
+    }
+
+    Stat cf_stat;
+    ret = StatParseResponse(buf, &cf_stat);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to parse the response from the server",
```
Comment (olehermanse) [inline libcfnet/protocol.c]: Fails here in travis:

```
Running cf-net:
==2682== Memcheck, a memory error detector
==2682== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==2682== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==2682== Command: /var/cfengine/bin/cf-net GET /var/cfengine/masterfiles/promises.cf
==2682== 
   error: ProtocolStat: Failed to parse the response from the server
Could not stat: '/var/cfengine/masterfiles/promises.cf'
==2682== 
==2682== HEAP SUMMARY:
==2682==     in use at exit: 112,688 bytes in 3,365 blocks
==2682==   total heap usage: 5,581 allocs, 2,216 frees, 486,238 bytes allocated
==2682== 
==2682== LEAK SUMMARY:
==2682==    definitely lost: 0 bytes in 0 blocks
==2682==    indirectly lost: 0 bytes in 0 blocks
==2682==      possibly lost: 0 bytes in 0 blocks
==2682==    still reachable: 112,688 bytes in 3,365 blocks
==2682==         suppressed: 0 bytes in 0 blocks
==2682== Reachable blocks (those to which a pointer was found) are not shown.
==2682== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2682== 
==2682== For counts of detected and suppressed errors, rerun with: -v
==2682== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Looking for problems in get.txt:
==2682== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
   error: ProtocolStat: Failed to parse the response from the server
   error: ProtocolStat: Failed to parse the response from the server
```

File: libcfnet/protocol.h (hunk 5 above)
Comment (karlhto) [reply libcfnet/protocol.h]: Nope, left over from its previous iteration. :^)

File: libcfnet/stat_cache.c (hunk 2 above)
Comment (karlhto) [reply libcfnet/stat_cache.c]: I think `0` should be fine, since it returns `-1` in any other case.

File: libcfnet/protocol.c (hunk 3 above)
Comment (karlhto) [reply libcfnet/protocol.c]: It isn't. And `time_t` should be safe until at least 2038 :^)

**Comment by karlhto** [comment]
@cf-bottom Jenkins pls

File: libcfnet/stat_cache.c [hunk 7]
```diff
@@ -208,144 +211,183 @@ int cf_remote_stat(AgentConnection *conn, bool encrypt, const char *file,
...
-        {
-            cfst.cf_readlink = xstrdup(recvbuffer + 3);
-        }
-        else
+    ret = StatParseResponse(recvbuffer, &cfst);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Cannot read SYNCH reply from '%s', only %d/13 items parsed",
+            conn->remoteip, ret);
```
Comment (vpodzime) [inline libcfnet/stat_cache.c]: This will print `-1/13 items parsed`.

File: libcfnet/stat_cache.c [hunk 8]
```diff
@@ -208,144 +211,183 @@ int cf_remote_stat(AgentConnection *conn, bool encrypt, const char *file,
...
+
+    // If remote path is symbolic link, receive actual path here
+    int recv_len = ReceiveTransaction(conn->conn_info, recvbuffer, NULL);
+    if (recv_len == -1)
+    {
+        /* TODO mark connection in the cache as closed. */
+        return -1;
+    }
+
+    if (recv_len > 4)
```
Comment (vpodzime) [inline libcfnet/stat_cache.c]: This `4` looks very arbitrary here.

File: libcfnet/stat_cache.c (hunk 2 above)
Comment (vpodzime) [reply libcfnet/stat_cache.c]: So there are only two options of what the function can return -- failure or success. Looks like a good candidate for `bool` return type to me.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4107/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4107/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4107/

File: libcfnet/stat_cache.c (hunk 7 above)
Comment (karlhto) [reply libcfnet/stat_cache.c]: Whoops, yeah. Thanks for noticing. : )

File: libcfnet/stat_cache.c (hunk 2 above)
Comment (karlhto) [reply libcfnet/stat_cache.c]: I agree : )

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4139/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4139/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4139/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4141/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4141/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4141/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4142/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4142/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4142/

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4154/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4154/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4154/

**Comment by vpodzime** [comment]
new jenkins build:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/4166/)

File: libcfnet/protocol.c [hunk 9]
```diff
@@ -197,14 +197,106 @@ bool ProtocolStatGet(AgentConnection *conn, const char *remote_path,
...
+
+    char buf[CF_BUFSIZE] = {0};
+    int to_send = snprintf(buf, CF_BUFSIZE, "SYNCH %jd STAT %s",
+                           (intmax_t) tloc, remote_path);
+
+    int ret = SendTransaction(conn->conn_info, buf, to_send, CF_DONE);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "Could not send stat request for remote file %s:%s.",
```
Comment (olehermanse) [inline libcfnet/protocol.c]: Log messages should not end with punctuation

File: libcfnet/protocol.c [hunk 10]
```diff
@@ -197,14 +197,106 @@ bool ProtocolStatGet(AgentConnection *conn, const char *remote_path,
...
+    mode_t file_type = FileTypeToMode(cf_stat.cf_type);
+    if (file_type == 0)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Invalid file type identifier for file %s:%s, %u",
+            conn->this_server, remote_path, cf_stat.cf_type);
+        return false;
+    }
+
+    stat_buf->st_mode = file_type | cf_stat.cf_mode;
```
Comment (olehermanse) [inline libcfnet/protocol.c]: Should explain what this does and why.

**Comment by karlhto** [comment]
@cf-bottom Jenkins

These failures seem unrelated, right?

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4186/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4186/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4186/

**Pull request merged**